### DISCUSSION TITLE
`ExpandingCutoffFilter` idea

### DISCUSSION THREAD

**Discussion by ninedigits** [alan-turing-institute/sktime#6327] [Ideas]
Is there interest in creating a hybrid of `ExpandingWindowSplitter` and `CutoffSplitter`? This concept stems from the need to backtest hierarchical data encompassing various products, each with a unique start date. The current `CutoffSplitter` targets specific dates but relies on a `window_length` parameter, making it resemble a `SlidingWindowSplitter` with a set cutoff rather than an initial window. Meanwhile, the `CutoffFhSplitter` shares a similar idea but does not support hierarchical data use.

The goal would be to:

1. Allow user to specify the cutoff dates
2. Use the full time series data up until the cutoff date
3. Support hierarchical data 
4. Integrate with existing infrastructure tools, like `evaluate` and `ForecastingGridSearchCV`

```
fh=3
step_size=1
...
s1: ---**********xxx---
s2: -************xxx---
s3: -------******xxx---

s1: ---***********xxx--
s2: -*************xxx--
s3: -------*******xxx--

...
```

**Comment by fkiraly** [comment]
Interesting! Sounds like a great idea!

I think it might be time to test our new splitter extension template and `check_estimator` integration!
https://github.com/sktime/sktime/blob/main/extension_templates/split.py

Let us know if you need help, but you've done this before (well, for other object types, but it should work similarly)

**Comment by ninedigits** [comment]
@fkiraly 
Follow-up question -- Consider a scenario where I need to forecast multiple products, each starting on a different date (ie some products are older than others and I want to use to the full time series available to me). I aim to identify the optimal combinations of parameters for each product using `ForecastByLevel` in conjunction with `ForecastingGridSearchCV`. Given that backtesting is computationally demanding and time-consuming, I prefer to execute it only when necessary. Once I determine the best parameters, I need to save them and reintroduce them to the model, ensuring that each specific time series is paired with its corresponding optimal parameters. Could the existing infrastructure support this approach? If so, how might the implementation look? I have an example provided below on how it might work with `ExpandingCutoffSplitter`.

```python3
forecaster = TransformedTargetForecaster(
    [
        ("deseasonalizer", OptionalPassthrough(Deseasonalizer(sp=4))),
        ("boxcox", OptionalPassthrough(BoxCoxTransformer())),
        ("detrender", OptionalPassthrough(Detrender())),
        ("forecaster",
            EnsembleForecaster(
                [
                    ("naive", NaiveForecaster()),
                    ("ets", StatsForecastAutoETS()),
                ]
            ),
        ),
    ]
)


param_grid = {
    "deseasonalizer__passthrough": [True, False],
    "boxcox__passthrough": [True, False],
    "detrender__passthrough": [True, False],
    "forecaster__naive__strategy": ["last", "mean", "drift"],
    "forecaster__ets__damped": [True, False],
}

cutoffs = [pd.Timestamp("2021-Q1"), pd.Timestamp('2021-Q2')]

cv = ExpandingCutoffSplitter(fh=[1, 2, 3], cutoffs=cutoffs)

gscv = ForecastByLevel(
    ForecastingGridSearchCV(
        forecaster=forecaster,
        param_grid=param_grid,
        cv=cv,
        scoring=MeanAbsolutePercentageError()
    )
)
gscv.fit(df)

### Want to be able to do the following
# 1) Get a list of the best possible params, currently available from ForecastingGridSearchCV
#    but not when wrapped in ForecastByLevel.
list_best_possible_params = gscv.best_params_ 

# 2) Take list of params and plug it back into real-time forecasts.
model = ForecastByLevel(
    forecaster,
    forecaster_set_method={
        "_set_params": list_best_possible_params
    }
)
model.fit(y_train, fh)
model.predict()```
